clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java

```diff
@@ -155,28 +150,140 @@ public VeniceChangelogConsumerImpl(
         storeName,
-        storeCurrentVersion,
         partitionCount,
         viewClassName);
   }
 
   @Override
   public CompletableFuture<Void> subscribe(Set<Integer> partitions) {
+    return internalSubscribe(partitions, null);
+  }
+
+  public CompletableFuture<Void> internalSubscribe(Set<Integer> partitions, String topic) {
     return CompletableFuture.supplyAsync(() -> {
       try {
-        readOnlySchemaRepository.start();
-        readOnlySchemaRepository.subscribe(storeName);
+        storeRepository.start();
+        storeRepository.subscribe(storeName);
       } catch (InterruptedException e) {
         throw new RuntimeException(e);
       }
-      readOnlySchemaRepository.refresh();
+      storeRepository.refresh();
+
+      String topicToSubscribe;
+      if (topic == null) {
+        topicToSubscribe = getCurrentServingVersionTopic();
+      } else {
+        topicToSubscribe = topic;
+      }
       Set<TopicPartition> topicPartitionSet = kafkaConsumer.assignment();
-      List<TopicPartition> topicPartitionList = getPartitionListToSubscribe(partitions, topicPartitionSet);
+      List<TopicPartition> topicPartitionList =
+          getPartitionListToSubscribe(partitions, topicPartitionSet, topicToSubscribe);
       kafkaConsumer.assign(topicPartitionList);
+      for (TopicPartition topicPartition: topicPartitionList) {
+        if (!topicPartition.topic().endsWith(ChangeCaptureView.CHANGE_CAPTURE_TOPIC_SUFFIX)) {
+          compressorMap.put(topicPartition.partition(), getVersionCompressor(topicPartition));
+        }
+      }
       kafkaConsumer.seekToBeginning(topicPartitionList);
-      subscribedPartitions.addAll(partitions);
+      return null;
+    });
+  }
+
+  protected VeniceCompressor getVersionCompressor(TopicPartition topicPartition) {
+    Store store = storeRepository.getStore(storeName);
+    Version version = store.getVersion(Version.parseVersionFromVersionTopicName(topicPartition.topic())).get();
+    VeniceCompressor compressor;
+    if (CompressionStrategy.ZSTD_WITH_DICT.equals(version.getCompressionStrategy())) {
+      compressor = compressorFactory.getVersionSpecificCompressor(topicPartition.topic());
+      if (compressor == null) {
+        // we need to retrieve the dictionary from the kafka topic
+        ByteBuffer dictionary = DictionaryUtils.readDictionaryFromKafka(
+            topicPartition.topic(),
+            new VeniceProperties(changelogClientConfig.getConsumerProperties()));
+        compressor = compressorFactory.createVersionSpecificCompressorIfNotExist(
+            version.getCompressionStrategy(),
+            topicPartition.topic(),
+            dictionary.array());
+      }
+    } else {
+      compressor = compressorFactory.getCompressor(version.getCompressionStrategy());
+    }
+    return compressor;
+  }
+
+  @Override
+  public CompletableFuture<Void> seekToBeginningOfPush(Set<Integer> partitions) {
+    // Get latest version topic
+    storeRepository.refresh();
```

Comment:
We can use the method of `getCurrentServingVersionTopic` to avoid duplication here.